### PR TITLE
Add VXLAN support (topic/jsiwek/vxlan)

### PR DESCRIPTION
**Pull request by jsiwek**
The `topic/jsiwek/vxlan` is in both `zeek` and `zeek-doc` repos.  Note that I preserved a squashed commit of the original contribution by Henrik, mostly as a reminder to give credit in the changelog, etc.

### CODE REVIEW

**Comment by jsiwek** [comment]
And just for reference, this fixes #250

File: src/analyzer/protocol/vxlan/VXLAN.cc [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		}
+
+	data += pkt.hdr_size;
+	len -= pkt.hdr_size;
+	caplen -= pkt.hdr_size;
+
+	IP_Hdr* inner = nullptr;
+	int res = 0;
+
+	switch ( pkt.l3_proto ) {
```
Comment (rsmmr) [inline src/analyzer/protocol/vxlan/VXLAN.cc]: Hmm, could we send this into `NetSessions::DoNextPacket` here? That would avoid hard-coding another l3_proto switch. Not sure if that works, just wondering.

File: src/analyzer/protocol/udp/UDP.cc [hunk 2]
```diff
@@ -61,7 +62,30 @@ void UDP_Analyzer::DeliverPacket(int len, const u_char* data, bool is_orig,
 
 	int chksum = up->uh_sum;
 
-	if ( ! ignore_checksums && caplen >= len )
+	auto validate_checksum = ! ignore_checksums && caplen >=len;
+	constexpr auto vxlan_len = 8;
+	constexpr auto eth_len = 14;
+
+	if ( validate_checksum &&
```
Comment (rsmmr) [inline src/analyzer/protocol/udp/UDP.cc]: The need for hard-coding VXLAN into UDP is unfortunate, but I don't see another way of doing it either, given the weird checksum semantics.

**Comment by rsmmr** [comment]
I got the merge ready to push, just going to wait for responses to the two questions.

File: scripts/base/init-bare.bro [hunk 3]
```diff
@@ -4857,6 +4857,18 @@ export {
...
+	## Whether to validate the checksum supplied in the outer UDP header
+	## of a VXLAN encapsulation.  The spec says the checksum should be
+	## transmitted as zero, but if not, then the decapsulating destination
+	## may choose whether to perform the validation.
+	const validate_vxlan_checksums = T &redef;
+
+	## The set of UDP ports used for VXLAN traffic.  Traffic using this
+	## UDP destination port will attempt to be decapsulated.  Note that if
+	## if you customize this, you may still want to manually ensure that
+	## :bro:see:`likely_server_ports` also gets populated accordingly.
```
[the remark this answers is not in the document]
Comment (jsiwek) [reply scripts/base/init-bare.bro]: Not that I can see due to parsing order -- `redef` will apply the default value of `vxlan_ports` to `likely_server_ports` first and then only afterward add a new value to `vxlan_ports`).  Example:

```
$ cat test.bro
redef Tunnel::vxlan_ports += { 1111/udp };

event bro_init()
	{
	print Tunnel::vxlan_ports;
	print 1111/udp in likely_server_ports;
	}

$ bro ./test.bro
{
1111/udp,
4789/udp
}
F
```

So `redef` here is maybe not the best interface for customizing `likely_server_ports`, maybe ought to have functions that modify it.

I notice a bug though: I forgot to add `&redef` to `vxlan_ports`, mind adding it during merge ?

File: src/analyzer/protocol/vxlan/VXLAN.cc (hunk 1 above)
Comment (jsiwek) [reply src/analyzer/protocol/vxlan/VXLAN.cc]: Except I think we do want control back here for the purpose of additional/custom error checking/handling.  E.g. don't want a generic Weird, but rather ProtocolViolation if there's a problem parsing that layer.  Also do want to possibly do more stuff in the case where that layer looks good.  E.g. ProtocolConfirmation and raise further vxlan-specific events.

We might probably still do all that if we further change `NetSessions::DoNextPacket` to allow for it: add return value, suppress/customize weird messages, etc., but seems too early or out of scope.  Other tunnel analysis also already goes through `DoNextInnerPacket`, so it's at least consistent.

File: src/analyzer/protocol/udp/UDP.cc (hunk 2 above)
Comment (jsiwek) [reply src/analyzer/protocol/udp/UDP.cc]: Same feeling, but couldn't think of a less-awkward way to handle it.

**Comment by jsiwek** [comment]
Thanks for the review, I think the one thing to fix is the missing `&redef` on `vxlan_ports`, can you get it?

File: scripts/base/init-bare.bro (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsmmr) [reply scripts/base/init-bare.bro]: Ok, I see. Will add the &redef.

File: src/analyzer/protocol/vxlan/VXLAN.cc (hunk 1 above)
Comment (rsmmr) [reply src/analyzer/protocol/vxlan/VXLAN.cc]: Ok, sounds like opportunity for a bit of refactoring in the future, possibly along with other L2 work.

**Pull request merged**